snow/networking/worker/pool.go

```diff
@@ -12,51 +16,81 @@ type Request func()
 type Pool interface {
 	// Send the request to the worker pool.
 	//
-	// Send should never be called after [Shutdown] is called.
+	// Send can be safely called after [Shutdown] and it'll be no-op.
 	Send(Request)
 
 	// Shutdown the worker pool.
 	//
 	// This method will block until all workers have finished their current
 	// tasks.
 	//
-	// It is safe to call shutdown multiple times.
+	// Shutdown can be safely called multiple times.
 	Shutdown()
 }
 
 type pool struct {
 	requests chan Request
 
+	// [noMoreSends] helps making Send request no-op
+	// is issued after Shutdown
+	noMoreSends utils.Atomic[bool]
+
+	// [shutdownOnce] ensures Shutdown idempotency
 	shutdownOnce sync.Once
-	shutdownWG   sync.WaitGroup
+
+	// [shutdownWG] makes sure all workers have stopped before Shutdown returns
+	shutdownWG sync.WaitGroup
+
+	// closing [quit] tells the workers to stop working
+	quit chan struct{}
 }
 
-func NewPool(size int) Pool {
+func NewPool(workersCount int) Pool {
 	p := &pool{
 		requests: make(chan Request),
+		quit:     make(chan struct{}),
 	}
-	p.shutdownWG.Add(size)
-	for w := 0; w < size; w++ {
+
+	for w := 0; w < workersCount; w++ {
+		p.shutdownWG.Add(1)
 		go p.runWorker()
 	}
+
 	return p
 }
 
 func (p *pool) runWorker() {
 	defer p.shutdownWG.Done()
 
-	for request := range p.requests {
-		request()
+	for {
+		select {
+		case <-p.quit:
+			return // stop worker
+		case request := <-p.requests:
+			if request != nil {
+				request()
+			}
+		}
+	}
+}
+
+func (p *pool) Send(msg Request) {
+	if p.noMoreSends.Get() {
+		return
+	}
+
```

Comment:
I don't follow what this is needed for. The request queue isn't buffered... so I don't really see what we are gaining out of having this flag at all.

Reply:
The idea is to avoid requests being executed after shutdown is called. Indeed this should be guaranteed by the fact that a closed channel should be always picked in a select. Dropped the flag.